src/UI/GameBuilder.java

```diff
@@ -28,27 +34,70 @@
 
   public GameInstance createGame(PlayerSetupInfo[] coInfos)
   {
-    GameScenario scenario = new GameScenario(unitModelScheme, incomePerCity, startingFunds, isFowEnabled);
+    GameScenario scenario = new GameScenario(unitModelScheme, incomePerCity, startingFunds, isFowEnabled, tagMode);
 
     // Add any CO-specific units into our set of UnitModels
     for(int i = 0; i < mapInfo.getNumCos(); ++i)
     {
       coInfos[i].getCurrentCO().injectUnits(unitModelScheme.getGameReadyModels());
     }
 
-    // Create all of the commanders.
-    Commander[] cos = new Commander[mapInfo.getNumCos()];
+    // Create all of the armies.
+    Army[] armies = new Army[mapInfo.getNumCos()];
     for(int i = 0; i < mapInfo.getNumCos(); ++i)
     {
-      cos[i] = coInfos[i].makeCommander(scenario.rules);
+      armies[i] = new Army();
+      // TODO: Add logic to add cart/persistent tags tag partners
+      armies[i].cos = new Commander[] { coInfos[i].makeCommander(scenario.rules) };
+      armies[i].team = coInfos[i].currentTeam;
+      armies[i].setAIController(coInfos[i].getCurrentAI().create(armies[i]));
     }
 
     // Build the CO list and the new map and create the game instance.
-    MapMaster map = new MapMaster( cos, mapInfo );
+    MapMaster map = new MapMaster( armies, mapInfo );
+
+    // If doing team merge... merge the teams
+    if( TagMode.Team_Merge == tagMode )
+    {
+      // The final number of armies should probably be fewer than the number of armies in the map, so we'll just copy them over
+      ArrayList<Army> finalArmies = new ArrayList<>();
+      for(Army army : armies)
+      {
+        // See if we've seen this army's team yet
+        Army match = null;
+        for (Army finalArmy : finalArmies)
+          if(finalArmy.team == army.team)
+            match = finalArmy;
+
+        // If not, just add it to the final list 
+        if(match == null)
+        {
+          finalArmies.add(army);
+        }
+        else // else, merge with the match we found
+        {
+          // credit: https://www.baeldung.com/java-concatenate-arrays
+          // Clone the final army's commander list into a larger space, then copy the new Commanders into it
+          Commander[] newCoList = Arrays.copyOf(match.cos, match.cos.length + army.cos.length);
+          System.arraycopy(army.cos, 0, newCoList, match.cos.length, army.cos.length);
+          match.cos = newCoList;
+          match.HQLocations.addAll(army.HQLocations);
+        }
```

Comment:
Why create each army with one CO and then merge instead of just creating each army with the full list of COs up front?

Reply:
Because the list of "people who own units" needs to be in order and the right size (and you can't just assume a jagged iteration of armies->COs will be right, since who's grouped is user-defined).
Then `HQLocations` moved to `Army` and I said "eh".
It's probably possible to have `MapMaster`'s constructor handle only COs and offer some API for pulling the HQs out, but I figure that's probably more work than what we have now.